src/widgettracker/index.ts

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) Jupyter Development Team.
+// Distributed under the terms of the Modified BSD License.
+
+import {
+  IDisposable, DisposableDelegate
+} from 'phosphor-disposable';
+
+import {
+  ISignal, Signal
+} from 'phosphor-signaling';
+
+import {
+  TabPanel
+} from 'phosphor-tabs';
+
+import {
+  Widget
+} from 'phosphor-widget';
+
+
+/**
+ * The class name added to the currently active widget's title.
+ */
+const SEMANTIC_FOCUS_CLASS = 'jp-mod-semanticFocus';
+
+
+/**
+ * An object that tracks the active widget in an application.
+ */
+export
+class WidgetTracker<T extends Widget> implements IDisposable {
+  /**
+   * Construct a new widget tracker.
+   */
+  constructor() {
+    // TODO: Replace this with message filters on semantic-focus
+    // events when available.
+    document.body.addEventListener('focus', this, true);
+  }
+
+  /**
+   * A signal emitted when the active widget changes.
+   */
+  get activeWidgetChanged(): ISignal<WidgetTracker<T>, T> {
+    return activeWidgetChangedSignal.bind(this);
+  }
+
+  /**
+   * Test whether the widget tracker has been disposed.
+   *
+   * #### Notes
+   * This is a read-only property.
+   */
+  get isDisposed(): boolean {
+    return this._widgets = null;
+  }
+
+  /**
+   * The read-only list of tracked widgets.
+   *
+   * #### Notes
+   * This is a read-only property.
+   */
+  get widgets(): T[] {
```

Comment:
Disposed widgets are automatically removed from the tracker. If the caller holds onto the returned array (a shallow copy), then yes.